homeassistant/components/fibaro/__init__.py

```diff
@@ -357,39 +374,78 @@ def _read_devices(self):
-        """Stop Fibaro Thread."""
-        _LOGGER.info("Shutting down Fibaro connection")
-        for controller in hass.data[FIBARO_CONTROLLERS].values():
-            controller.disable_state_handler()
+    # check if already configured
+    if hass.config_entries.async_entries(DOMAIN):
+        return True
+
+    for gateway in gateways:
+        # prepare new config based on configuration.yaml
+        conf = {
+            CONF_URL: gateway[CONF_URL],
+            CONF_USERNAME: gateway[CONF_USERNAME],
+            CONF_PASSWORD: gateway[CONF_PASSWORD],
+            CONF_IMPORT_PLUGINS: gateway[CONF_PLUGINS],
+        }
+
+        # import into config flow based configuration
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+            )
+        )
 
-    hass.data[FIBARO_DEVICES] = {}
+    return True
+
+
+def _init_controller(data: dict[str, Any]) -> FibaroController:
+    """Validate the user input allows us to connect to fibaro."""
+    controller = FibaroController(data)
+    controller.connect_with_error_handling()
+    return controller
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the Fibaro Component."""
+    try:
+        controller = await hass.async_add_executor_job(_init_controller, entry.data)
+    except FibaroConnectFailed as connect_ex:
+        raise ConfigEntryNotReady(
+            f"Could not connect to controller at {entry.data[CONF_URL]}"
+        ) from connect_ex
+    except FibaroAuthFailed:
+        return False
+
+    data: dict[str, Any] = {}
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data
+    data[FIBARO_CONTROLLER] = controller
+    devices = data[FIBARO_DEVICES] = {}
     for platform in PLATFORMS:
-        hass.data[FIBARO_DEVICES][platform] = []
+        devices[platform] = [*controller.fibaro_devices[platform]]
 
-    for gateway in gateways:
-        controller = FibaroController(gateway)
-        if controller.connect():
-            hass.data[FIBARO_CONTROLLERS][controller.hub_serial] = controller
-            for platform in PLATFORMS:
-                hass.data[FIBARO_DEVICES][platform].extend(
-                    controller.fibaro_devices[platform]
-                )
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
 
-    if hass.data[FIBARO_CONTROLLERS]:
-        for platform in PLATFORMS:
-            discovery.load_platform(hass, platform, DOMAIN, {}, base_config)
-        for controller in hass.data[FIBARO_CONTROLLERS].values():
-            controller.enable_state_handler()
-        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_fibaro)
-        return True
+    controller.enable_state_handler()
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    _LOGGER.info("Shutting down Fibaro connection")
```

Comment:
Does this need to be `info` level message? Seems a bit high...

Reply:
Good catch 👍 

This should be max debug.

Reply:
I created PR #68783 to change the log level.

Reply:
Thanks!